python/ray/serve/model_wrappers.py

```diff
@@ -204,22 +196,38 @@ async def predict_impl(inp: Union[np.ndarray, "pd.DataFrame"]):
 
             @serve.batch(**batching_params)
             async def predict_impl(inp: Union[List[np.ndarray], List["pd.DataFrame"]]):
-
+                batch_size = len(inp)
                 if isinstance(inp[0], np.ndarray):
-                    collate_func = collate_array
+                    batched = BatchingManager.batch_array(inp)
                 elif pd is not None and isinstance(inp[0], pd.DataFrame):
-                    collate_func = collate_dataframe
+                    batched = BatchingManager.batch_dataframe(inp)
+                elif isinstance(inp[0], dict):
+                    batched = BatchingManager.batch_dict_array(inp)
                 else:
                     raise ValueError(
-                        "ModelWrapper only accepts numpy array or dataframe as input "
+                        "ModelWrapper only accepts numpy array, dataframe, or dict of "
+                        "array as input "
                         f"but got types {[type(i) for i in inp]}"
                     )
 
-                batched, unpack = collate_func(inp)
                 out = self.model.predict(batched, **predict_kwargs)
                 if isinstance(out, ray.ObjectRef):
                     out = await out
-                return unpack(out)
+
+                if isinstance(out, np.ndarray):
+                    return BatchingManager.split_array(out, batch_size)
+                elif pd is not None and isinstance(out, pd.DataFrame):
+                    return BatchingManager.split_dataframe(out, batch_size)
+                elif isinstance(out, dict):
+                    return BatchingManager.split_dict_array(out, batch_size)
+                elif isinstance(out, list) and len(out) == batch_size:
+                    return out
+                else:
+                    raise ValueError(
+                        f"ModelWrapper only accepts list of length {batch_size}, numpy "
```

Comment:
This error might be a bit confusing if a too-short or too-long list is passed in since the type will still be `list`. Could we refactor the message to also include a length if the input type is a list?